### DISCUSSION TITLE
How do I set the ItemsPanelTemplate content in code-behind C#?

### DISCUSSION THREAD

**Discussion by ekral** [AvaloniaUI/Avalonia#13556] [Q&A] [answered]
I have this xaml code:

```xaml
<ListBox Name="listBox">
	<ListBox.ItemsPanel>
		<ItemsPanelTemplate>
				<WrapPanel/>
		</ItemsPanelTemplate>
	</ListBox.ItemsPanel>
</ListBox>
```

And I would like to define the ListBox in the code-behind in C# only. I know how to [create a DataTemplate in code](https://docs.avaloniaui.net/docs/templates/creating-data-templates-in-code), but how do I create a WrapPanel instance and assign it to the ItemsPanelTemplate Content property?

The following code does not work:

```csharp
 ListBox listBox = new ListBox()
 {
     ItemsPanel = new ItemsPanelTemplate()
     {
         Content = new WrapPanel() // Error 
     }
 };
```

**Comment by workgroupengineering** [comment, marked as answer]
https://docs.avaloniaui.net/docs/next/concepts/templates/creating-data-templates-in-code

https://github.com/AvaloniaUI/Avalonia/blob/9cd8dda5206ffbcb6705ada2ae8d3753aaaf9caf/tests/Avalonia.Controls.UnitTests/ListBoxTests.cs#L792-L818

https://github.com/AvaloniaUI/Avalonia/blob/9cd8dda5206ffbcb6705ada2ae8d3753aaaf9caf/tests/Avalonia.LeakTests/ControlTests.cs#L882-L896

**Comment by ekral** [comment]
Thank you, it works:

```csharp
ListBox listBox = new ListBox()
{
    ItemsPanel = new FuncTemplate<Panel?>(() => new WrapPanel())
};
```

Avalonia is great, but it still needs better documentation. I think I will try to add to the docs myself.
